charms/reactive/endpoints.py

```diff
@@ -341,6 +341,8 @@ def __init__(self, relation_id):
         self._units = None
         self._departed_units = None
         self._data = None
+        # to check relation has application relation data
+        self._app = False
```

Comment:
Instead of marking the entire relation as app vs unit, it would better to just have separate data buckets. There are actually potential use-cases for using both app and unit level data on the relation. This is also missing a way to read remote app data.

```suggestion
        self._app_data = None
        self._remote_app_data = None
```